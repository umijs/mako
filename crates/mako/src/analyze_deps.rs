use swc_css_ast::{ImportHref, Url, UrlValue};
use swc_css_visit::VisitWith as CSSVisitWith;
use swc_ecma_ast::{CallExpr, Callee, Decl, Expr, Import, Lit, ModuleDecl, ModuleItem, Stmt};
use swc_ecma_visit::{Visit, VisitWith};

use crate::module::{Dependency, ModuleAst, ResolveType};

pub fn analyze_deps(ast: &ModuleAst) -> Vec<Dependency> {
    match ast {
        ModuleAst::Script(ast) => analyze_deps_js(ast),
        ModuleAst::Css(ast) => analyze_deps_css(ast),
        _ => {
            vec![]
        }
    }
}

pub fn add_swc_helper_deps(deps: &mut Vec<Dependency>, ast: &ModuleAst) {
    match ast {
        ModuleAst::Script(ast) => {
            ast.body.iter().for_each(|stmt| {
                // var x = require('x'); -> x
                if let ModuleItem::Stmt(Stmt::Decl(Decl::Var(box decl))) = stmt {
                    let x = &decl.decls[0].init.as_ref();
                    if x.is_none() {
                        return;
                    }
                    if let box Expr::Call(call_expr) = x.unwrap() {
                        if is_commonjs_require(&call_expr) {
                            if let Some(src) = get_first_arg_str(&call_expr) {
                                if src.starts_with("@swc/helpers") {
                                    deps.push(Dependency {
                                        source: src,
                                        resolve_type: ResolveType::Require,
                                        // why 0?
                                        // swc helpers are always inserted before other modules
                                        order: 0 as usize,
                                    });
                                }
                            }
                        }
                    }
                }
            });
        }
        _ => {}
    }
}

fn analyze_deps_js(ast: &swc_ecma_ast::Module) -> Vec<Dependency> {
    let mut visitor = DepCollectVisitor::new();
    ast.visit_with(&mut visitor);
    visitor.dependencies
}

fn analyze_deps_css(ast: &swc_css_ast::Stylesheet) -> Vec<Dependency> {
    let mut visitor = DepCollectVisitor::new();
    ast.visit_with(&mut visitor);
    visitor.dependencies
}

struct DepCollectVisitor {
    dependencies: Vec<Dependency>,
    dep_strs: Vec<String>,
    order: usize,
}

impl DepCollectVisitor {
    fn new() -> Self {
        Self {
            dependencies: vec![],
            dep_strs: vec![],
            // start with 1
            // 0 for swc helpers
            order: 1,
        }
    }
    fn bind_dependency(&mut self, source: String, resolve_type: ResolveType) {
        if !self.dep_strs.contains(&source) {
            self.dep_strs.push(source.clone());
            self.dependencies.push(Dependency {
                source,
                order: self.order,
                resolve_type,
            });
            self.order += 1;
        }
    }
}

impl Visit for DepCollectVisitor {
    fn visit_module_decl(&mut self, n: &ModuleDecl) {
        match n {
            ModuleDecl::Import(import) => {
                if import.type_only {
                    return;
                }
                let src = import.src.value.to_string();
                self.bind_dependency(src, ResolveType::Import);
            }
            ModuleDecl::ExportNamed(export) => {
                if let Some(src) = &export.src {
                    let src = src.value.to_string();
                    self.bind_dependency(src, ResolveType::ExportNamed);
                }
            }
            ModuleDecl::ExportAll(export) => {
                let src = export.src.value.to_string();
                self.bind_dependency(src, ResolveType::ExportAll);
            }
            _ => {}
        }
        // export function xxx() {} 里可能包含 require 或 import()
        n.visit_children_with(self);
    }
    fn visit_call_expr(&mut self, expr: &CallExpr) {
        if is_commonjs_require(expr) {
            if let Some(src) = get_first_arg_str(expr) {
                self.bind_dependency(src, ResolveType::Require);
                return;
            }
        } else if is_dynamic_import(expr) {
            if let Some(src) = get_first_arg_str(expr) {
                self.bind_dependency(src, ResolveType::DynamicImport);
                return;
            }
        }
        expr.visit_children_with(self);
    }
}

impl swc_css_visit::Visit for DepCollectVisitor {
    fn visit_import_href(&mut self, n: &ImportHref) {
        match n {
            // e.g.
            // @import url(a.css)
            // @import url("a.css")
            ImportHref::Url(url) => {
                let src: Option<String> = url.value.as_ref().map(|box value| match value {
                    UrlValue::Str(str) => str.value.to_string(),
                    UrlValue::Raw(raw) => raw.value.to_string(),
                });
                if let Some(src) = src {
                    self.bind_dependency(src, ResolveType::Css);
                }
            }
            // e.g.
            // @import "a.css"
            ImportHref::Str(src) => {
                let src = src.value.to_string();
                self.bind_dependency(src, ResolveType::Css);
            }
        }
        // remove visit children since it is not used currently
        // n.visit_children_with(self);
    }

    fn visit_url(&mut self, n: &Url) {
        // 检查 url()
        let href_string = n
            .value
            .as_ref()
            .map(|box value| match value {
                UrlValue::Str(str) => str.value.to_string(),
                UrlValue::Raw(raw) => raw.value.to_string(),
            })
            .unwrap();
        self.bind_dependency(href_string, ResolveType::Css);
        // n.visit_children_with(self);
    }
}

pub fn is_dynamic_import(call_expr: &CallExpr) -> bool {
    matches!(&call_expr.callee, Callee::Import(Import { .. }))
}

pub fn is_commonjs_require(call_expr: &CallExpr) -> bool {
    if let Callee::Expr(box Expr::Ident(swc_ecma_ast::Ident { sym, .. })) = &call_expr.callee {
        sym == "require"
    } else {
        false
    }
}

fn get_first_arg_str(call_expr: &CallExpr) -> Option<String> {
    if let Some(arg) = call_expr.args.first() {
        if let box Expr::Lit(Lit::Str(str_)) = &arg.expr {
            return Some(str_.value.to_string());
        }
    }
    None
}

#[cfg(test)]
mod tests {
    use std::{
        collections::HashMap,
        path::PathBuf,
        sync::{Arc, Mutex, RwLock},
    };

    use crate::{
        ast::build_js_ast,
        chunk_graph::ChunkGraph,
        compiler::{Context, Meta},
        config::Config,
        module_graph::ModuleGraph,
    };

    use super::{add_swc_helper_deps, analyze_deps_js};

    #[test]
    fn test_analyze_deps() {
        let deps = resolve(
            r#"
import 'foo';
            "#
            .trim(),
            false,
        );
        assert_eq!(deps, "foo");
    }

    #[test]
    fn test_analyze_deps_inside_exports() {
        let deps = resolve(
            r#"
export function foo() {
    require('foo');
}           
            "#
            .trim(),
            false,
        );
        assert_eq!(deps, "foo");
    }

    #[test]
    fn test_analyze_deps_ignore_type_only() {
        let deps = resolve(
            r#"
import type { x } from 'foo';
import 'bar';
            "#
            .trim(),
            false,
        );
        assert_eq!(deps, "bar");
    }

    #[test]
    fn test_add_swc_helper_deps() {
        let deps = resolve(
            r#"
var a = require("@swc/helpers/a");
var b = require("foo");
var c = require("@swc/helpers/b");
            "#
            .trim(),
            true,
        );
        assert_eq!(deps, "@swc/helpers/a,@swc/helpers/b");
    }

    fn resolve(code: &str, swc_helper: bool) -> String {
        let root = PathBuf::from("/path/to/root");
        let ast = build_js_ast(
            "test.js",
            code,
            &Arc::new(Context {
                config: Config::new(&root).unwrap(),
                root,
                module_graph: RwLock::new(ModuleGraph::new()),
                chunk_graph: RwLock::new(ChunkGraph::new()),
                assets_info: Mutex::new(HashMap::new()),
                meta: Meta::new(),
            }),
        );
        let mut deps = vec![];
        if swc_helper {
            let ast = crate::module::ModuleAst::Script(ast);
            add_swc_helper_deps(&mut deps, &ast);
        } else {
            deps.extend(analyze_deps_js(&ast));
        }
        let deps = deps
            .iter()
            .map(|dep| dep.source.as_str())
            .collect::<Vec<_>>()
            .join(",");
        deps
    }
}
