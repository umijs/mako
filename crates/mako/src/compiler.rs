use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;
use std::sync::{Arc, Mutex, RwLock};
use std::time::Instant;

use anyhow::{anyhow, Error, Result};
use colored::Colorize;
use regex::Regex;
use swc_core::common::sync::Lrc;
use swc_core::common::{Globals, SourceMap, DUMMY_SP};
use swc_core::ecma::ast::Ident;
use swc_node_comments::SwcComments;
use tracing::debug;

use crate::ast::comments::Comments;
use crate::ast::file::win_path;
use crate::config::{Config, Mode, ModuleIdStrategy, OutputMode};
use crate::generate::chunk_graph::ChunkGraph;
use crate::generate::optimize_chunk::OptimizeChunksInfo;
use crate::module_graph::ModuleGraph;
use crate::plugin::{Plugin, PluginDriver, PluginGenerateEndParams};
use crate::plugins;
use crate::resolve::{get_resolvers, Resolvers};
use crate::share::helpers::SWC_HELPERS;
use crate::stats::StatsInfo;
use crate::utils::id_helper::{assign_numeric_ids, compare_modules_by_incoming_edges};
use crate::utils::{thread_pool, ParseRegex};

pub struct Context {
    pub module_graph: RwLock<ModuleGraph>,
    pub chunk_graph: RwLock<ChunkGraph>,
    pub assets_info: Mutex<HashMap<String, String>>,
    pub modules_with_missing_deps: RwLock<Vec<String>>,
    pub config: Config,
    pub numeric_ids_map: RwLock<HashMap<String, usize>>,
    pub args: Args,
    pub root: PathBuf,
    pub meta: Meta,
    pub plugin_driver: PluginDriver,
    pub stats_info: StatsInfo,
    pub resolvers: Resolvers,
    pub static_cache: RwLock<MemoryChunkFileCache>,
    pub optimize_infos: Mutex<Option<Vec<OptimizeChunksInfo>>>,
}

#[derive(Default)]
pub struct MemoryChunkFileCache {
    content_map: HashMap<String, (Vec<u8>, u64)>,
    root: Option<PathBuf>,
}

impl MemoryChunkFileCache {
    pub fn new(root: Option<PathBuf>) -> Self {
        Self {
            content_map: HashMap::new(),
            root,
        }
    }

    pub fn write<T: AsRef<str>>(&mut self, path: T, content: Vec<u8>, hash: u64) -> Result<()> {
        let str = path.as_ref();

        if let Some((_, in_mem_hash)) = self.content_map.get(str) {
            if *in_mem_hash != hash {
                self.write_to_disk(str, &content)?;
            }
        } else {
            self.write_to_disk(str, &content)?;
        }
        self.content_map
            .insert(path.as_ref().to_string(), (content, hash));
        Ok(())
    }

    pub fn read<T: AsRef<str>>(&self, path: T) -> Option<Vec<u8>> {
        self.content_map
            .get(path.as_ref())
            .map(|(content, _)| content.clone())
    }

    fn write_to_disk<T: AsRef<str>>(&self, path: T, content: &[u8]) -> Result<()> {
        if let Some(root) = &self.root {
            let path = root.join(path.as_ref());
            fs::write(path, content)?;
        }
        Ok(())
    }
}

#[derive(Default)]
pub struct Args {
    pub watch: bool,
}

impl Context {
    pub fn write_static_content<T: AsRef<str>>(
        &self,
        path: T,
        content: Vec<u8>,
        hash: u64,
    ) -> Result<()> {
        let mut map = self.static_cache.write().unwrap();
        map.write(path, content, hash)
    }

    pub fn get_static_content<T: AsRef<str>>(&self, path: T) -> Option<Vec<u8>> {
        let map = self.static_cache.read().unwrap();
        map.read(path)
    }
}

impl Default for Context {
    fn default() -> Self {
        let mut numeric_ids_map = HashMap::new();
        SWC_HELPERS.iter().enumerate().for_each(|(i, item)| {
            numeric_ids_map.insert(item.to_string(), i);
        });
        let config: Config = Default::default();
        let resolvers = get_resolvers(&config);
        Self {
            config,
            args: Args { watch: false },
            root: PathBuf::from(""),
            module_graph: RwLock::new(ModuleGraph::new()),
            chunk_graph: RwLock::new(ChunkGraph::new()),
            assets_info: Mutex::new(HashMap::new()),
            modules_with_missing_deps: RwLock::new(Vec::new()),
            meta: Meta::new(),
            plugin_driver: Default::default(),
            stats_info: StatsInfo::new(),
            resolvers,
            optimize_infos: Mutex::new(None),
            static_cache: Default::default(),
            numeric_ids_map: RwLock::new(numeric_ids_map),
        }
    }
}

pub struct Meta {
    pub script: ScriptMeta,
    pub css: CssMeta,
}

impl Meta {
    pub fn new() -> Self {
        Self {
            script: ScriptMeta::new(),
            css: CssMeta::new(),
        }
    }
}

impl Default for Meta {
    fn default() -> Self {
        Self::new()
    }
}

pub struct ScriptMeta {
    pub cm: Lrc<SourceMap>,
    pub origin_comments: RwLock<Comments>,
    pub output_comments: RwLock<Comments>,
    pub globals: Globals,
    // These idents may be used in other places, such as transform_async_module
    pub module_ident: Ident,
    pub exports_ident: Ident,
    pub require_ident: Ident,
}

impl ScriptMeta {
    fn new() -> Self {
        Self {
            cm: Default::default(),
            origin_comments: Default::default(),
            output_comments: Default::default(),
            globals: Globals::default(),
            module_ident: build_ident("module"),
            exports_ident: build_ident("exports"),
            require_ident: build_ident("__mako_require__"),
        }
    }
}

fn build_ident(ident: &str) -> Ident {
    Ident {
        ctxt: Default::default(),
        span: DUMMY_SP,
        sym: ident.into(),
        optional: false,
    }
}

pub struct CssMeta {
    pub cm: Lrc<SourceMap>,
    pub globals: Globals,
    pub comments: SwcComments,
}

impl CssMeta {
    fn new() -> Self {
        Self {
            cm: Default::default(),
            globals: Globals::default(),
            comments: Default::default(),
        }
    }
}

impl Context {
    pub fn emit_assets(&self, origin_path: String, output_path: String) {
        let mut assets_info = self.assets_info.lock().unwrap();
        assets_info.insert(origin_path, output_path);
    }
}

pub struct Compiler {
    pub context: Arc<Context>,
}

impl Compiler {
    pub fn new(
        config: Config,
        root: PathBuf,
        args: Args,
        extra_plugins: Option<Vec<Arc<dyn Plugin>>>,
    ) -> Result<Self> {
        if !root.is_absolute() {
            return Err(anyhow!("root path must be absolute"));
        }

        let root = PathBuf::from(win_path(root.to_str().unwrap()));

        // why add plugins before builtin plugins?
        // because plugins like less-loader need to be added before assets plugin
        // TODO: support plugin orders
        let mut plugins: Vec<Arc<dyn Plugin>> = vec![];
        if let Some(extra_plugins) = extra_plugins {
            plugins.extend(extra_plugins);
        }
        let builtin_plugins: Vec<Arc<dyn Plugin>> = vec![
            // features
            Arc::new(plugins::manifest::ManifestPlugin {}),
            Arc::new(plugins::copy::CopyPlugin {}),
            Arc::new(plugins::import::ImportPlugin {}),
            // file types
            Arc::new(plugins::context_module::ContextModulePlugin {}),
            Arc::new(plugins::runtime::MakoRuntime {}),
            Arc::new(plugins::invalid_webpack_syntax::InvalidWebpackSyntaxPlugin {}),
            Arc::new(plugins::hmr_runtime::HMRRuntimePlugin {}),
            Arc::new(plugins::wasm_runtime::WasmRuntimePlugin {}),
            Arc::new(plugins::async_runtime::AsyncRuntimePlugin {}),
            Arc::new(plugins::emotion::EmotionPlugin {}),
            Arc::new(plugins::tree_shaking::FarmTreeShake {}),
            Arc::new(plugins::detect_circular_dependence::LoopDetector {}),
        ];
        plugins.extend(builtin_plugins);

        let mut config = config;

        if config.mode == Mode::Production && config.experimental.imports_checker {
            plugins.push(Arc::new(plugins::imports_checker::ImportsChecker {}));
        }

        if let Some(progress) = &config.progress {
            plugins.push(Arc::new(plugins::progress::ProgressPlugin::new(
                plugins::progress::ProgressPluginOptions {
                    prefix: "Mako".to_string(),
                    template:
                        "‚óè {prefix:.bold} {bar:25.green/white.dim} ({percent}%) {wide_msg:.dim}"
                            .to_string(),
                    progress_chars: progress.progress_chars.clone(),
                },
            )));
        }

        if let Some(duplicate_package_checker) = &config.check_duplicate_package {
            plugins.push(Arc::new(
                plugins::duplicate_package_checker::DuplicatePackageCheckerPlugin::new()
                    .show_help(duplicate_package_checker.show_help)
                    .emit_error(duplicate_package_checker.emit_error)
                    .verbose(duplicate_package_checker.verbose),
            ));
        }

        if config.experimental.require_context {
            plugins.push(Arc::new(plugins::require_context::RequireContextPlugin {}))
        }

        if config.output.mode == OutputMode::Bundless {
            plugins.insert(
                0,
                Arc::new(plugins::bundless_compiler::BundlessCompilerPlugin {}),
            );
        }

        if std::env::var("DEBUG_GRAPH").is_ok_and(|v| v == "true") {
            plugins.push(Arc::new(plugins::graphviz::Graphviz {}));
        }

        if args.watch && std::env::var("SSU").is_ok_and(|v| v == "true") {
            plugins.push(Arc::new(plugins::ssu::SUPlus::new()));
        }

        if args.watch && config.experimental.central_ensure {
            plugins.push(Arc::new(plugins::central_ensure::CentralChunkEnsure {}));
        }

        if let Some(minifish_config) = &config._minifish {
            let inject = if let Some(inject) = &minifish_config.inject {
                let mut map = HashMap::new();

                for (k, ii) in inject.iter() {
                    map.insert(
                        k.clone(),
                        plugins::minifish::Inject {
                            from: ii.from.clone(),
                            name: k.clone(),
                            named: ii.named.clone(),
                            namespace: ii.namespace,
                            exclude: ii.exclude.parse_into_regex()?,
                            include: ii.include.parse_into_regex()?,
                            prefer_require: ii.prefer_require.map_or(false, |v| v),
                        },
                    );
                }
                Some(map)
            } else {
                None
            };

            plugins.insert(
                0,
                Arc::new(plugins::minifish::MinifishPlugin {
                    mapping: minifish_config.mapping.clone(),
                    meta_path: minifish_config.meta_path.clone(),
                    inject,
                }),
            );
        }

        let ignores = config
            .ignores
            .iter()
            .map(|ignore| Regex::new(ignore).map_err(Error::new))
            .collect::<Result<Vec<Regex>>>()?;
        plugins.push(Arc::new(plugins::ignore::IgnorePlugin { ignores }));

        let plugin_driver = PluginDriver::new(plugins);

        plugin_driver.modify_config(&mut config, &root, &args)?;

        let resolvers = get_resolvers(&config);
        let mut numeric_ids_map = HashMap::new();
        SWC_HELPERS.iter().enumerate().for_each(|(i, item)| {
            numeric_ids_map.insert(item.to_string(), i);
        });
        Ok(Self {
            context: Arc::new(Context {
                static_cache: if config.write_to_disk {
                    RwLock::new(MemoryChunkFileCache::new(Some(config.output.path.clone())))
                } else {
                    Default::default()
                },
                config,
                args,
                root,
                module_graph: RwLock::new(ModuleGraph::new()),
                chunk_graph: RwLock::new(ChunkGraph::new()),
                assets_info: Mutex::new(HashMap::new()),
                modules_with_missing_deps: RwLock::new(Vec::new()),
                meta: Meta::new(),
                plugin_driver,
                numeric_ids_map: RwLock::new(numeric_ids_map),
                stats_info: StatsInfo::new(),
                resolvers,
                optimize_infos: Mutex::new(None),
            }),
        })
    }

    pub fn compile(&self) -> Result<()> {
        // ÂÖàÊ∏ÖÁ©∫ dist ÁõÆÂΩï
        if self.context.config.clean {
            self.clean_dist()?;
        }

        let t_compiler = Instant::now();
        let start_time = chrono::Local::now().timestamp_millis();
        let building_with_message = format!(
            "Building with {} for {}...",
            "mako".to_string().cyan(),
            self.context.config.mode
        )
        .green();
        println!("{}", building_with_message);
        {
            crate::mako_profile_scope!("Build Stage");
            let files = self
                .context
                .config
                .entry
                .values()
                .map(|entry| {
                    let mut entry = entry.to_string_lossy().to_string();
                    let is_browser = matches!(
                        self.context.config.platform,
                        crate::config::Platform::Browser
                    );
                    let watch = self.context.args.watch;
                    let hmr = self.context.config.hmr.is_some();
                    if is_browser && watch && hmr {
                        entry = format!("{}?hmr", entry);
                    }
                    crate::ast::file::File::new_entry(entry, self.context.clone())
                })
                .collect();
            self.context.plugin_driver.build_start(&self.context)?;

            self.build(files)?;

            debug!("start after build");

            self.context
                .plugin_driver
                .after_build(&self.context, self)?;
        }

        self.context.plugin_driver.before_generate(&self.context)?;

        if let ModuleIdStrategy::Numeric = self.context.config.module_id_strategy {
            let module_graph = self.context.module_graph.read().unwrap();
            assign_numeric_ids(
                module_graph.modules(),
                |a, b| compare_modules_by_incoming_edges(&module_graph, &a.id, &b.id),
                |module, id| {
                    let mut numeric_ids_map = self.context.numeric_ids_map.write().unwrap();
                    // reserved ten indexes for swc helper and others runtime module
                    numeric_ids_map.insert(module.id.id.clone(), id + 10);
                },
            )
        }

        let result = {
            crate::mako_profile_scope!("Generate Stage");
            // need to put all rayon parallel iterators run in the existed scope, or else rayon
            // will create a new thread pool for those parallel iterators
            thread_pool::scope(|_| self.generate())
        };
        let t_compiler_duration = t_compiler.elapsed();
        match result {
            Ok(mut stats) => {
                stats.start_time = start_time;
                stats.end_time = chrono::Local::now().timestamp_millis();
                println!(
                    "{}",
                    format!(
                        "‚úì Built in {}",
                        format!("{}ms", t_compiler_duration.as_millis()).bold()
                    )
                    .green()
                );
                if !self.context.args.watch {
                    println!("{}", "Complete!".bold());
                }
                let params = PluginGenerateEndParams {
                    is_first_compile: true,
                    time: t_compiler.elapsed().as_millis() as i64,
                    stats,
                };
                self.context
                    .plugin_driver
                    .generate_end(&params, &self.context)?;
                self.context.plugin_driver.write_bundle(&self.context)?;
                Ok(())
            }
            Err(e) => Err(e),
        }
    }

    pub fn full_hash(&self) -> u64 {
        crate::mako_profile_function!();
        let cg = self.context.chunk_graph.read().unwrap();
        let mg = self.context.module_graph.read().unwrap();
        cg.full_hash(&mg)
    }

    fn clean_dist(&self) -> Result<()> {
        // compiler ÂâçÊ∏ÖÈô§ distÔºåÂ¶ÇÊûúÂêéÁª≠ dev ÁéØÂ¢É‰∏çÂú® output_path ÈáåÔºåÈúÄË¶ÅÂÜçË°•‰∏ä dev ÁöÑÈÄªËæë
        let output_path = &self.context.config.output.path;
        if fs::metadata(output_path).is_ok() {
            fs::remove_dir_all(output_path)?;
        }
        Ok(())
    }
}
