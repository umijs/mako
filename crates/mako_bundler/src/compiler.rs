use crate::{
    build::build::BuildParam, config::Config, context::Context, generate::generate::GenerateParam,
};

pub struct Compiler {
    pub context: Context,
}

impl Compiler {
    pub fn new(config: Config) -> Self {
        let context = Context::new(config);
        Self { context }
    }

    pub fn run(&mut self) {
        self.build(&BuildParam { files: None });
        self.generate(&GenerateParam { write: true });
    }

    pub fn _watch(&self) {}
}
