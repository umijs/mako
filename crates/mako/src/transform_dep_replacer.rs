use std::collections::HashMap;
use std::sync::Arc;

use swc_atoms::js_word;
use swc_common::DUMMY_SP;
use swc_ecma_ast::Expr::Call;
use swc_ecma_ast::{
    AssignExpr, AssignOp, BindingIdent, BlockStmt, CallExpr, Callee, Decl, Expr, ExprOrSpread,
    ExprStmt, FnExpr, Function, Ident, Lit, MemberExpr, MemberProp, NamedExport, NewExpr, Pat,
    PatOrExpr, Stmt, Str, ThrowStmt, VarDecl, VarDeclKind, VarDeclarator,
};
use swc_ecma_utils::quote_ident;
use swc_ecma_visit::{VisitMut, VisitMutWith};

use crate::analyze_deps::{is_commonjs_require, is_dynamic_import};
use crate::compiler::Context;
use crate::module::Dependency;

pub struct DepReplacer<'a> {
    pub to_replace: &'a DependenciesToReplace,
    pub context: &'a Arc<Context>,
}

#[derive(Debug, Clone)]
pub struct DependenciesToReplace {
    pub resolved: HashMap<String, String>,
    pub missing: HashMap<String, Dependency>,
}

fn miss_throw_stmt<T: AsRef<str>>(source: T) -> Expr {
    let fn_expr = Expr::Fn(FnExpr {
        ident: Some(quote_ident!("makoMissingModule")),
        function: Box::new(Function {
            is_async: false,
            params: vec![],
            decorators: vec![],
            span: DUMMY_SP,
            body: Some(BlockStmt {
                span: DUMMY_SP,
                stmts: vec![
                    Stmt::Decl(Decl::Var(Box::new(VarDecl {
                        span: DUMMY_SP,
                        kind: VarDeclKind::Var,
                        declare: false,
                        decls: vec![VarDeclarator {
                            span: DUMMY_SP,
                            name: Pat::Ident(BindingIdent {
                                id: quote_ident!("e"),
                                type_ann: None,
                            }),
                            init: Some(Box::new(Expr::New(NewExpr {
                                span: DUMMY_SP,
                                callee: Box::new(Expr::Ident(Ident::new(
                                    js_word!("Error"),
                                    DUMMY_SP,
                                ))),
                                args: Some(vec![ExprOrSpread {
                                    spread: None,
                                    expr: Box::new(Expr::Lit(Lit::Str(Str {
                                        span: DUMMY_SP,
                                        value: format!("Cannot find module '{}'", source.as_ref())
                                            .into(),
                                        raw: None,
                                    }))),
                                }]),
                                type_args: None,
                            }))),
                            definite: false,
                        }],
                    }))),
                    Stmt::Expr(ExprStmt {
                        span: DUMMY_SP,
                        expr: Box::new(Expr::Assign(AssignExpr {
                            span: DUMMY_SP,
                            left: PatOrExpr::Expr(Box::new(Expr::Member(MemberExpr {
                                span: DUMMY_SP,
                                obj: Box::new(Expr::Ident(quote_ident!("e"))),
                                prop: MemberProp::Ident(quote_ident!("code")),
                            }))),
                            op: AssignOp::Assign,
                            right: Box::new(Expr::Lit(Lit::Str(Str {
                                span: DUMMY_SP,
                                value: "MODULE_NOT_FOUND".into(),
                                raw: None,
                            }))),
                        })),
                    }),
                    Stmt::Throw(ThrowStmt {
                        span: DUMMY_SP,
                        arg: Box::new(Expr::Ident(quote_ident!("e"))),
                    }),
                ],
            }),
            return_type: None,
            type_params: None,
            is_generator: false,
        }),
    });

    Call(CallExpr {
        span: DUMMY_SP,
        callee: Callee::Expr(Box::new(Expr::Ident(quote_ident!("Object")))),
        args: vec![ExprOrSpread {
            spread: None,
            expr: Box::new(Call(CallExpr {
                span: DUMMY_SP,
                args: vec![],
                callee: Callee::Expr(Box::new(fn_expr)),
                type_args: None,
            })),
        }],
        type_args: None,
    })
}

impl VisitMut for DepReplacer<'_> {
    fn visit_mut_expr(&mut self, expr: &mut Expr) {
        if let Expr::Call(call_expr) = expr {
            if is_commonjs_require(call_expr) || is_dynamic_import(call_expr) {
                if let ExprOrSpread {
                    expr: box Expr::Lit(Lit::Str(ref mut source)),
                    ..
                } = &mut call_expr.args[0]
                {
                    let source_string = source.value.clone().to_string();

                    match self.to_replace.missing.get(&source_string) {
                        Some(_) => {
                            call_expr.args[0] = ExprOrSpread {
                                spread: None,
                                expr: Box::new(miss_throw_stmt(&source_string)),
                            };
                        }
                        None => {
                            self.replace_source(source);
                        }
                    }
                }
            }
        }
        expr.visit_mut_children_with(self);
    }

    fn visit_mut_import_decl(&mut self, import_decl: &mut swc_ecma_ast::ImportDecl) {
        self.replace_source(&mut import_decl.src);
    }

    fn visit_mut_named_export(&mut self, n: &mut NamedExport) {
        if let Some(ref mut src) = n.src {
            self.replace_source(src.as_mut());
        }
    }
}

impl DepReplacer<'_> {
    fn replace_source(&mut self, source: &mut Str) {
        if let Some(replacement) = self.to_replace.resolved.get(&source.value.to_string()) {
            let span = source.span;

            let module_id_string = replacement.clone();
            //generate_module_id(replacement.clone(), self.context);

            // NOTE: JsWord 有缓存，直接设置 value 的方式在这种情况下不会生效
            // if (process.env.NODE_ENV === 'development') { require("./foo") }
            *source = Str::from(module_id_string);
            // 保持原来的 span，不确定不加的话会不会导致 sourcemap 错误

            source.span = span;
        }
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashMap;
    use std::sync::Arc;

    use maplit::hashmap;
    use swc_common::GLOBALS;
    use swc_ecma_visit::VisitMut;

    use crate::assert_display_snapshot;
    use crate::ast::build_js_ast;
    use crate::compiler::Context;
    use crate::module::{Dependency, ResolveType};
    use crate::test_helper::transform_ast_with;
    use crate::transform_dep_replacer::{DepReplacer, DependenciesToReplace};

    #[test]
    fn test_simple_replace() {
        let context: Arc<Context> = Arc::new(Default::default());

        GLOBALS.set(&context.meta.script.globals, || {
            let mut ast = build_js_ast("index.jsx",
                                       r#"require("react")"#
                                       , &context.clone()).unwrap();

            let to_replace  = DependenciesToReplace {
                resolved: hashmap! {"react".to_string()=> "/root/node_modules/react/index.js".to_string()},
                missing: HashMap::new(),
            };

            let cloned = context.clone();
            let mut visitor: Box<dyn VisitMut> = Box::new(DepReplacer {
                to_replace: &to_replace,
                context: &cloned,
            });

            assert_display_snapshot!(transform_ast_with(&mut ast.ast, &mut visitor, &context.meta.script.cm));
        });
    }

    #[test]
    fn test_replace_missing_dep() {
        let context: Arc<Context> = Arc::new(Default::default());

        GLOBALS.set(&context.meta.script.globals, || {
            let mut ast =
                build_js_ast("index.jsx", r#"require("react")"#, &context.clone()).unwrap();

            let to_replace = DependenciesToReplace {
                resolved: HashMap::new(),
                missing: hashmap! {"react".to_string() => Dependency {
                    resolve_type: ResolveType::Import,
                    source: "react".to_string(),
                    span: None,
                    order: 0,
                }},
            };

            let cloned = context.clone();
            let mut visitor: Box<dyn VisitMut> = Box::new(DepReplacer {
                to_replace: &to_replace,
                context: &cloned,
            });

            assert_display_snapshot!(transform_ast_with(
                &mut ast.ast,
                &mut visitor,
                &context.meta.script.cm
            ));
        });
    }

    #[test]
    fn test_replace_top_level_missing_dep_in_try() {
        let context: Arc<Context> = Arc::new(Default::default());

        GLOBALS.set(&context.meta.script.globals, || {
            let mut ast = build_js_ast(
                "index.jsx",
                r#"
                                       try {require("react")}catch(e){}"#,
                &context.clone(),
            )
            .unwrap();

            let to_replace = DependenciesToReplace {
                resolved: HashMap::new(),
                missing: hashmap! {"react".to_string() => Dependency {
                    resolve_type: ResolveType::Import,
                    source: "react".to_string(),
                    span: None,
                    order: 0,
                }},
            };

            let cloned = context.clone();
            let mut visitor: Box<dyn VisitMut> = Box::new(DepReplacer {
                to_replace: &to_replace,
                context: &cloned,
            });

            assert_display_snapshot!(transform_ast_with(
                &mut ast.ast,
                &mut visitor,
                &context.meta.script.cm
            ));
        });
    }

    #[test]
    fn test_import_replace() {
        assert_display_snapshot!(transform_code("import x from 'x'"));
    }

    #[test]
    fn test_export_from_replace() {
        assert_display_snapshot!(transform_code("export {x} from 'x'"));
    }

    #[test]
    fn test_dynamic_import_from_replace() {
        assert_display_snapshot!(transform_code("const x = import('x')"));
    }

    fn transform_code(code: &str) -> String {
        let context: Arc<Context> = Arc::new(Default::default());

        GLOBALS.set(&context.meta.script.globals, || {
            let mut ast = build_js_ast("test.js", code, &context).unwrap();

            let mut visitor = DepReplacer {
                to_replace: &DependenciesToReplace {
                    resolved: hashmap! {
                        "x".to_string() => "/x/index.js".to_string()
                    },
                    missing: hashmap! {},
                },
                context: &context,
            };

            transform_ast_with(&mut ast.ast, &mut visitor, &context.meta.script.cm)
        })
    }
}
