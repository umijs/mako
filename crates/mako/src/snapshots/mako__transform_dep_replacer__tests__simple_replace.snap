---
source: crates/mako/src/transform_dep_replacer.rs
expression: "transform_ast_with(&mut ast.ast, &mut visitor)"
---
require("/root/node_modules/react/index.js");
